### PR TITLE
Add a SoyMapWrites type-class along with helper functions.

### PR DESCRIPTION
**Pull request by ClaireNeveu**
@pozsi This PR creates a new `SoyMapWrites` type-class along with utility functions for using and constructing the type-class.

This change allows users to preserve type information when converting objects to Soy as they are usually converted to `SoyMap` which has extra functionality over other `SoyValue`s.

[Trello ticket](https://trello.com/c/OwFDvnAK/83-create-a-soymapwrites-type-class)

### CODE REVIEW

File: src/main/scala/com/kinja/soy/SoyWrites.scala [hunk 1]
```diff
@@ -21,6 +21,25 @@ trait SoyWrites[-T] extends DefaultSoyWrites {
 object SoyWrites extends DefaultSoyWrites
 
 /**
+ * The class of types that can be converted into a SoyMap. Every SoyMapWrites implies a SoyWrites.
+ */
+trait SoyMapWrites[-T] extends SoyWrites[T] {
+  def toSoy(t: T): SoyMap
+  final def and[U](w: SoyMapWrites[U]): SoyMapWrites[(T, U)] = {
+    val _toSoy = toSoy _
```
Comment (pjrt) [inline src/main/scala/com/kinja/soy/SoyWrites.scala]: That's weird. Why is that necessary? Scala issue?

Comment (ClaireNeveu) [inline src/main/scala/com/kinja/soy/SoyWrites.scala]: It will be shadowed by the definition of `toSoy` below.

Comment (pjrt) [inline src/main/scala/com/kinja/soy/SoyWrites.scala]: Ah, I see. You could also add a self-type variable to the top SoyMapWrites and get around that. IE

`trait SoyMapWrites[-T] { a: =>`

and then 

`a.toSoy`

**Comment by pozsi** [comment]
LGTM
